packages/devextreme/js/__internal/ui/speech_to_text/speech_to_text.ts

```diff
@@ -0,0 +1,261 @@
+    super._init();
+    this._handleCustomEngineState();
+    this._createActions();
+  }
+
+  _initMarkup(): void {
+    super._initMarkup();
+    this.$element().addClass(SPEECH_TO_TEXT_CLASS);
+    this._renderButton();
+    this._updateButtonState();
+  }
+
+  private _createActions(): void {
+    ACTIONS.forEach((action) => {
+      this._setAction(action);
+    });
+  }
+
+  private _setAction(action: keyof SpeechToTextActions): void {
+    this._actions[action] = this._createActionByOption(action, {
+      excludeValidators: ['disabled', 'readOnly'],
+    }) || noop;
+  }
+
+  private _renderButton(): void {
+    this._button = this._createComponent<Button, ButtonProperties>(
+      this.$element(),
+      Button,
+      this._getButtonOptions(),
+    );
+  }
+
+  private _getButtonOptions(): ButtonProperties {
+    const {
+      activeStateEnabled,
+      focusStateEnabled,
+      hoverStateEnabled,
+      stylingMode,
+      type,
+      disabled,
+      width,
+      height,
+    } = this.option();
+
+    return {
+      stylingMode,
+      type,
+      disabled,
+      width,
+      height,
+      activeStateEnabled,
+      focusStateEnabled,
+      hoverStateEnabled,
+      icon: this._getCurrentIcon(),
+      text: this._getCurrentText(),
+      onClick: (e: ClickEvent): void => {
+        this._buttonClickHandler(e);
+      },
+    };
+  }
+
+  private _getCurrentIcon(): string | undefined {
+    const { startIcon, stopIcon } = this.option();
+
+    return this._isListening() ? stopIcon : startIcon;
+  }
+
+  private _getCurrentText(): string {
+    const { startText, stopText } = this.option();
+
+    return this._isListening()
+      ? stopText ?? ''
+      : startText ?? '';
+  }
+
+  private _buttonClickHandler(e: ClickEvent): void {
+    if (this._state === SpeechToTextState.DISABLED) {
+      return;
+    }
+
+    if (this._isListening()) {
+      this._handleStopClick(e);
+    } else {
+      this._handleStartClick(e);
+    }
+  }
+
+  private _handleStartClick(e: ClickEvent): void {
+    const { customSpeechRecognizer } = this.option();
+
+    if (!customSpeechRecognizer?.enabled) {
+      this._setState(SpeechToTextState.LISTENING);
+    }
+
+    this._actions.onStartClick?.({
+      component: this,
+      element: this.$element().get(0) as HTMLElement,
+      event: e.event,
+    });
+  }
+
+  private _handleStopClick(e: ClickEvent): void {
+    const { customSpeechRecognizer } = this.option();
+
+    if (!customSpeechRecognizer?.enabled) {
+      this._setState(SpeechToTextState.INITIAL);
+    }
+
+    this._actions.onStopClick?.({
+      component: this,
+      element: this.$element().get(0) as HTMLElement,
+      event: e.event,
+    });
+  }
+
+  private _setState(newState: SpeechToTextState): void {
+    if (this._state === newState) {
+      return;
+    }
+
+    this._state = newState;
+    this._updateButtonState();
+    this._updateCssClasses();
+  }
+
+  private _updateButtonState(): void {
```

Comment:
Let's make the naming more strict? It does not look like we gonna extend this method
```suggestion
  private _updateButtonIconAndText(): void {
```

Reply:
I don't think this is a good idea, because then depending on the state we will update the `aria-label` directly in this method.